File: Projeto-MobiFix/Database/dtos/serviceDtos/servicoDto.js
const paraServicoDto = (s) => {
    if (!s) return null;
    return {
        ServicoID: s._id,
        TrotineteNumSerie: s.trotineteId,
        Estado: s.estado,
        DataAgendamento: s.dataAgendamento,
        DescricaoDiagnostico: s.descricaoDiagnostico || null,
        FeedbackCliente: s.feedbackCliente || null,
        DataConclusao: s.dataConclusao || null,
        Preco: s.preco,
        HistoricoIntervencoes: (s.historicoIntervencoes || []).map(i => ({
            IntervencaoCatalogoID: i.intervencaoCatalogoId,
            MecanicoNumero: i.mecanicoId,
            DataInicio: i.dataInicio,
            DataFim: i.dataFim || null,
            TempoGastoMinutos: i.tempoGastoMinutos || null,
            PecasUtilizadas: (i.pecasUtilizadas || []).map(p => ({
                PecaEAN: p.pecaId,
                Quantidade: p.quantidade
            }))
        }))
    };
};

module.exports = { paraServicoDto };

File: Projeto-MobiFix/Backend/Models/Servicos/ServicoDto.cs
namespace Backend.Models;

using System.ComponentModel.DataAnnotations;

public class ServicoDto
{
    public int ServicoID { get; set; }
    public string TrotineteNumSerie { get; set; } = string.Empty;
    public string Estado { get; set; } = string.Empty;
    public DateTime DataAgendamento { get; set; }
    public string? DescricaoDiagnostico { get; set; }
    public string? FeedbackCliente { get; set; }
    public DateTime? DataConclusao { get; set; }
    public decimal Preco { get; set; }
    public List<IntervencaoRealizadaDto> HistoricoIntervencoes { get; set; } = new();
}

public class IntervencaoRealizadaDto
{
    public int IntervencaoCatalogoID { get; set; }
    public string MecanicoNumero { get; set; } = string.Empty;
    public DateTime DataInicio { get; set; }
    public DateTime? DataFim { get; set; }
    public int? TempoGastoMinutos { get; set; }
    public List<PecaUtilizadaDto> PecasUtilizadas { get; set; } = new();
}

public class PecaUtilizadaDto
{
    public string PecaEAN { get; set; } = string.Empty;
    public int Quantidade { get; set; }
}

public class ServicoCriacaoDto
{
    // ServicoID removido — é gerado pelo Node.js/MongoDB, não pelo cliente
    [Required(ErrorMessage = "O número de série da trotinete é obrigatório.")]
    public string TrotineteNumSerie { get; set; } = string.Empty;

    public string? FeedbackCliente { get; set; }
}

public class ServicoAtualizacaoDto
{
    public string? Estado { get; set; }
    public string? DescricaoDiagnostico { get; set; }
    public decimal? Preco { get; set; }
    public DateTime? DataConclusao { get; set; }
    public List<IntervencaoRealizadaDto>? HistoricoIntervencoes { get; set; }
}

public class LevantarTrotineteDto
{
    [Required]
    public string MetodoPagamento { get; set; } = string.Empty;
}

public class LevantamentoComFaturaDto
{
    public FaturaDto Fatura { get; set; } = null!;
}

public class TrotineteProntaDto
{
    public int ServicoID { get; set; }
    public string TrotineteNumSerie { get; set; } = string.Empty;
    public string? Marca { get; set; }
    public string? Modelo { get; set; }
    public string ClienteNIF { get; set; } = string.Empty;
    public DateTime DataAgendamento { get; set; }
    public DateTime? DataConclusao { get; set; }
    public decimal Preco { get; set; }
    public string? DescricaoDiagnostico { get; set; }

    // Mão de obra (= Preco do serviço) e detalhe das peças usadas
    // permitem ao operador ver o discriminado no checkout.
    public decimal MaoDeObra { get; set; }
    public decimal TotalPecas { get; set; }
    public decimal TotalFinal { get; set; }
    public List<PecaFaturacaoDto> Pecas { get; set; } = new();
}

public class PecaFaturacaoDto
{
    public string PecaEAN { get; set; } = string.Empty;
    public string Nome { get; set; } = string.Empty;
    public int Quantidade { get; set; }
    public decimal PrecoUnitario { get; set; }
}